BUG: Order of OS detection macros (for Windows) seems faulty

### Describe the issue:

The order of these OS detection macros might make Win64 and MinGW unreachable.

https://github.com/numpy/numpy/blob/07db5becf57373b0962483ecaf3452b075183e1f/numpy/core/include/numpy/npy_os.h#L22-L27

According to Microsoft [documentation](https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170) the macros are defined as follows:

>> _WIN32 Defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64. Otherwise, undefined.
>> _WIN64 Defined as 1 when the compilation target is 64-bit ARM or x64. Otherwise, undefined.

Essentially `_WIN32` is defined on 32 bit as well as 64 bit setups. `_WIN64` is defined on 64 bit Windows only. This means the first check 
```cpp
#elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32) 
     #define NPY_OS_WIN32 
```

will succeed and the numpy OS will be defined as `NPY_OS_WIN32` on 64-bit ARM or x64 Windows.

On top of that, MinGW compilers follow the same rule. So `_WIN32` is defined, say on 32 bit as well as 64 bit `MinGW64` `gcc/clang` while `_WIN64` is defined for 64 bit compilations only. 

Although I can not find a definite documentation for the MinGW64 compiler macros, as far as I can tell this is the state of these macros:

    64 bit MinGW compilers define `_MINGW32__`, `__MINGW64__`, `_WIN32`, and `_WIN64`.
    32 bit MinGW compilers define `__MINGW32__`, and `_WIN32`.
    64 bit MSVC compilation defines `_WIN32` and  `_WIN64`.
    32 bit MSVC compilation defines `_WIN32`.

Therefore, on 32 bit or 64 bit Windows or MinGW, the OS check will always say that the OS is `NPY_OS_WIN32`. This might be fine for most use cases (especially since the macros do not seem to be used in many places) but it is also not what is intended. 


---------------------------------------------------


With MinGW's 64-bit UCRT64 gcc, for example, this causes compilation failure because the OS is defined as `NPY_OS_WIN32` and this causes this check    

```cpp
#ifndef F2PY_THREAD_LOCAL_DECL
#if defined(_MSC_VER)
#define F2PY_THREAD_LOCAL_DECL __declspec(thread)
#elif defined(NPY_OS_MINGW)
#define F2PY_THREAD_LOCAL_DECL __thread
#elif defined(__STDC_VERSION__) \
      && (__STDC_VERSION__ >= 201112L) \
      && !defined(__STDC_NO_THREADS__) \
      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \
      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
   of `threads.h` when using an older release of glibc 2.12
   See gh-19437 for details on OpenBSD */
#include <threads.h>
#define F2PY_THREAD_LOCAL_DECL thread_local
#elif defined(__GNUC__) \
      && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4)))
#define F2PY_THREAD_LOCAL_DECL __thread
#endif
#endif
```

**in the generated files when using f2py** to assume `threads.h` is available. Compilation fails because it is not available.
  


### Reproduce the code example:

```python
Unfortunately I do not have a simple example to reproduce this. I am also not very familiar with numpy or f2py as much as I would like. I came across it while fixing things in a bigger project. 

On the other hand it should be easier to reproduce for someone who know what they are doing and have access to MYSY's UCRT64 environment. Even if that is not possible one can possible figure out what the results might be by looking at the relevant section in npy_os.h and the reported predefined macro status.
```


### Error message:

```sh
msys/tmp/tmp1rhc293t/src.mingw_x86_64_ucrt-3.11/curvif_simplifiedmodule.c:87:10: fatal error: threads.h: No such file or directory
   87 | #include <threads.h>
      |          ^~~~~~~~~~~
compilation terminated.
```

### Runtime information:

```sh
$> f2py

Version:     1.25.2
numpy Version: 1.25.2
Requires:    Python 3.5 or higher.
License:     NumPy license (see LICENSE.txt in the NumPy source code)
Copyright 1999 - 2011 Pearu Peterson all rights reserved.
https://web.archive.org/web/20140822061353/http://cens.ioc.ee/projects/f2py2e
```

### Context for the issue:

_No response_

--- a/numpy/core/include/numpy/npy_os.h
+++ b/numpy/core/include/numpy/npy_os.h
@@ -19,12 +19,12 @@
     #define NPY_OS_SOLARIS
 #elif defined(__CYGWIN__)
     #define NPY_OS_CYGWIN
-#elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32)
-    #define NPY_OS_WIN32
-#elif defined(_WIN64) || defined(__WIN64__) || defined(WIN64)
-    #define NPY_OS_WIN64
 #elif defined(__MINGW32__) || defined(__MINGW64__)
     #define NPY_OS_MINGW
+#elif defined(_WIN64) || defined(__WIN64__) || defined(WIN64)
+    #define NPY_OS_WIN64
+#elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32)
+    #define NPY_OS_WIN32
 #elif defined(__APPLE__)
     #define NPY_OS_DARWIN
 #elif defined(__HAIKU__)

--- a/numpy/core/include/numpy/npy_os.h
+++ b/numpy/core/include/numpy/npy_os.h
@@ -20,12 +20,12 @@
 #elif defined(__CYGWIN__)
     #define NPY_OS_CYGWIN
 /* We are on Windows.*/
-#elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32)
+#elif defined(_WIN32)
   /* We are using MinGW (64-bit or 32-bit)*/
   #if defined(__MINGW32__) || defined(__MINGW64__)
     #define NPY_OS_MINGW
   /* Otherwise, if _WIN64 is defined, we are targeting 64-bit Windows*/
-  #elif defined(_WIN64) || defined(__WIN64__) || defined(WIN64)
+  #elif defined(_WIN64)
     #define NPY_OS_WIN64
   /* Otherwise assume we are targeting 32-bit Windows*/
   #else

--- a/numpy/core/include/numpy/npy_os.h
+++ b/numpy/core/include/numpy/npy_os.h
@@ -19,12 +19,18 @@
     #define NPY_OS_SOLARIS
 #elif defined(__CYGWIN__)
     #define NPY_OS_CYGWIN
-#elif defined(__MINGW32__) || defined(__MINGW64__)
-    #define NPY_OS_MINGW
-#elif defined(_WIN64) || defined(__WIN64__) || defined(WIN64)
-    #define NPY_OS_WIN64
+/* We are on Windows.*/
 #elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32)
+  /* We are using MinGW (64-bit or 32-bit)*/
+  #if defined(__MINGW32__) || defined(__MINGW64__)
+    #define NPY_OS_MINGW
+  /* Otherwise, if _WIN64 is defined, we are targeting 64-bit Windows*/
+  #elif defined(_WIN64) || defined(__WIN64__) || defined(WIN64)
+    #define NPY_OS_WIN64
+  /* Otherwise assume we are targeting 32-bit Windows*/
+  #else
     #define NPY_OS_WIN32
+  #endif
 #elif defined(__APPLE__)
     #define NPY_OS_DARWIN
 #elif defined(__HAIKU__)

